### DISCUSSION TITLE
Version 4 implementations

### DISCUSSION THREAD

**Discussion by rhysmorgan134** [rhysmorgan134/node-CarPlay#61] [Show and tell]
@gozmanyoni Finally got an electron build done, it runs 100x times better than before, also tested with the PiMost and works amazingly! Just need to tie carplay started to a stream command via MOST, and figure how to make a navigatable app and still keep audio but video only when needed

<img width="1710" alt="image" src="https://github.com/rhysmorgan134/node-CarPlay/assets/36861722/52e4c130-cbe7-471c-9da3-90ca92ef53fd">

**Comment by gozmanyoni** [comment]
Amazing progress!

For a navigatable app - you can now subscribe to the command event from the dongle, and  the command "requestHostUI" fires when a user clicks on the Car logo in carplay - based on that you could switch to a different interface in the electron app.

Since we know the user switched away from the UI we can set a state variable and stop feeding jmuxer with frames? (that should continue audio, but stop decoding video?)

**Comment by rhysmorgan134** [comment]
@gozmanyoni PiMost react-carplay up and working! It auto switches amplifier to the pimost on plugged event, microphone next.

<img width="1342" alt="Screenshot 2023-10-02 at 20 45 49" src="https://github.com/rhysmorgan134/node-CarPlay/assets/36861722/e7d0f1b2-42b4-4bec-bfd6-72bffe29091e">

**Comment by gozmanyoni** [comment]
Amazing progress!

Had a thought - on my car for instance, it can take over the sink again if i activate the voice assistant, or press radio and so on - We could detect this and take over again (The way the MOST MMI does it, is by emulating a CD changer, then you can connect/disconnect from the cars interface by playing a CD)

**Reply by rhysmorgan134** [reply]
Interesting, I have actually been playing with this a small bit, I was hoping to not have to remove a device to emulate, but the issue I think is going to be that we will have no concept of whether another device has taken over the sink from the PiMost. If we emulate a device and remove that device from the ring, then yes we should get all disconnect notifications etc when something else takes over. I think this may be down to manufacturer implementation.

On my Land Rover, there is no notification when a source is changed, however on my XF (same hardware) I do get a notification that could likely be used. Another possibility for my cars atleast, is that all source switching is handled via the PiMost, which I think is perfectly do-able.

I'm still waiting to get my hands on a microphone, and I'll implement the mic next!

**Reply by rhysmorgan134** [reply]
To be fair I have an old accessory interface in the jag, which contains an old ipod connector. I could in theory take this modules identity and not lose anything!

**Reply by gozmanyoni** [reply]
That's a good candidate in the Jag - can you specifically select it from the cars interface?, in the bmw I disconnected the cd changer, it's not really practical to have one in this era 😁 I can still use CD's in the main head unit , but only one at a time

**Reply by rhysmorgan134** [reply]
yeah you can not only select it, but also browse through ipod folders, playlists etc. I am hoping it follows the MOST spec for an Auxillary FBlock, then I should be able to also feed the media data from carplay as current track etc!

The CD player on both of mine is integrated along with the radio and aux in, so far from ideal!
